velox/expression/tests/ExprOptimizerTest.cpp

```diff
@@ -0,0 +1,497 @@
+
+  void SetUp() override {
+    functions::prestosql::registerAllScalarFunctions("");
+    parse::registerTypeResolver();
+    options_.parseDecimalAsDouble = false;
+  }
+
+  void TearDown() override {
+    expression::ExprRewriteRegistry::instance().clear();
+  }
+
+  /// Validates the result of optimizing expression of given type matches the
+  /// expected expression.
+  /// @param expression Input expression to be optimized.
+  /// @param expected Expected expression after optimization.
+  /// @param type Type of input and expected expressions.
+  void testExpression(
+      const core::TypedExprPtr& expression,
+      const core::TypedExprPtr& expected,
+      const RowTypePtr& type) {
+    auto optimized =
+        expression::optimize(expression, queryCtx_.get(), pool(), false);
+    SCOPED_TRACE(fmt::format(
+        "Input: {}\nOptimized: {}\nExpected: {}",
+        expression->toString(),
+        optimized->toString(),
+        expected->toString()));
+    ASSERT_TRUE(*optimized == *expected);
+
+    const auto data = fuzzFlat(type);
+    const auto optimizedResult = evaluate(optimized, data);
+    const auto expectedResult = evaluate(expected, data);
+    test::assertEqualVectors(optimizedResult, expectedResult);
+  }
+
+  /// Validates expression optimization and provides an option to specify input
+  /// types for lambda expressions in the input and expected expressions.
+  /// @param input Input expression to be optimized.
+  /// @param expected Expected expression after optimization.
+  /// @param type Type of input and expected expressions.
+  /// @param lambdaInputTypes Types of inputs to lambda expression that could be
+  /// present in 'input'.
+  /// @param expectedLambdaInputTypes Types of inputs to lambda expression that
+  /// could be present in 'expected'.
+  void testExpression(
+      const std::string& input,
+      const std::string& expected,
+      const RowTypePtr& type = ROW({}),
+      const std::vector<TypePtr>& lambdaInputTypes = {},
+      const std::vector<TypePtr>& expectedLambdaInputTypes = {}) {
+    const auto expr = makeTypedExpr(input, type, lambdaInputTypes);
+    const auto expectedExpr =
+        makeTypedExpr(expected, type, expectedLambdaInputTypes);
+    testExpression(expr, expectedExpr, type);
+  }
+
+  /// Validates that the subexpressions that throw an exception during
+  /// optimization are replaced with a fail expression when optimize is called
+  /// with 'replaceEvalErrorWithFailExpr = true'. Also validates that the
+  /// optimized expression containing fail expression(s) will throw the expected
+  /// error upon evaluation.
+  /// @param input Input expression to be optimized.
+  /// @param expected Expected expression as a string.
+  /// @param error Expected error message upon evaluating input expression.
+  /// @param type Type of input expression.
+  /// @param lambdaInputTypes Types of inputs to lambda expression that could be
+  /// present in 'input'.
+  void assertFailCall(
```

Comment:
Can we combine assertFailCall and testExpression ? We should have one function which processes input/expected similarly and validates expected based on error

Its parameters would be 
{
      const std::string& input,
      const std::string& expected,
      const RowTypePtr& type = ROW({}),
      const std::vector<TypePtr>& lambdaInputTypes = {},
      const std::vector<TypePtr>& expectedLambdaInputTypes = {},
      bool error
}